feat: add page for environment variables (#29)
tuono-labs/tuono-documentation: src/routes/documentation/configuring/environment-variables.mdx

import MetaTags from '@/components/MetaTags'

<MetaTags

>   title="Tuono - Environment Variables"

Reviewer: ```suggestion
  title="Tuono - Environment Variables"
  description="Learn to add and access environment variables in your Tuono application"
```

  canonical="https://tuono.dev/documentation/environment-variables"
/>

import Breadcrumbs from '@/components/Breadcrumbs'

<Breadcrumbs breadcrumbs={[{ label: 'Environment Variables' }]} />

# Environment Variables

## Overview

> Tuono supports environment variables on both client and server, providing hot reload in both scenarios.
> 
> - Loads from `.env.*` files
> - Manages precedence across multiple env files
> - Supports hot reload on both client and server
> - Bundles client env vars with the `TUONO_PUBLIC_` prefix

Reviewer 2: The features list should include an introduction:

```suggestion
Tuono supports environment variables for both the client and server, 
with hot reload functionality in each scenario. 

Key features include:

- Loading from `.env.*` files
- Managing precedence across multiple env files
- Supporting hot reload on both client and server
- Bundling client env vars with the `TUONO_PUBLIC_` prefix
```

## Usage

> Define your environment variables in `.env` file(s). Here we will use a `.env` and a `.env.local`.

Reviewer 2: Rephrased to be more fluent

```suggestion
Define your environment variables in `.env` file(s). 
For example, you can use a `.env` file and a `.env.local` file.
```

```env
# .env
MY_KEY=foo
MY_DUPLICATED_KEY=env
MY_QUOTED_KEY="bar"
TUONO_PUBLIC_MY_PUBLIC_KEY=public_value
```

```env
# .env.local
MY_DUPLICATED_KEY=local
MY_SECRET=some_secret
```

> Here, variables set in `.env.local` will override variables set in `.env`. We cover this in more detail in the [modes section](#mode-specific-env-files).

Reviewer 2: I think we should be more concise here.

```suggestion
Variables set in `.env.local` will override those in `.env`. 
For more details, refer to the [modes section](#mode-specific-env-files).
```

> ### Clientside

Reviewer 2: ```suggestion
### Client-side
```

> Tuono makes any environment variable starting with the `TUONO_PUBLIC_` prefix available to the client on the `import.meta.env` object. Here's a usage example:

Reviewer 2: ```suggestion
Tuono exposes any environment variable that starts with the `TUONO_PUBLIC_` prefix to the client via the `import.meta.env` object. 
Here’s an example of how to use it:
```

```ts
console.log(import.meta.env.TUONO_PUBLIC_SOME_KEY) // "123"
console.log(import.meta.env.DB_PASSWORD) // undefined
```

> Note that variables starting with the `TUONO_PUBLIC_` prefix will end up in your client bundle - therefore these variables should **not** contain any sensitive or secret values.

Reviewer 2: This is an important information so it should be highlighted

```suggestion
> ⚠️ Note that variables starting with the TUONO_PUBLIC_ prefix will be included in your client bundle. 
> Therefore, these variables should not contain any sensitive or secret values.
```

Reviewer: Looks like this explanation is already been done few sentences above. I'd avoid repetitions. I think it is better putting this quote in the `client-side` section

Author: I think this is needed since it's a note on security, and isn't it in the client-side section already?

Reviewer: Yeah I agree! Better to keep it

> ### Serverside

Reviewer 2: ```suggestion
### Server-side
```

> Environment variables are automatically loaded into the rust environment, allowing you to use them just like you would use a system environment variable in any other rust codebase.

Reviewer 2: ```suggestion
Environment variables are automatically loaded into the Rust environment, 
allowing you to use them just like any other system environment variable in a Rust codebase.
```

```rs
use std::env;
use tuono_lib::Request;
use tuono_lib::axum::http::StatusCode;

#[tuono_lib::api(GET)]
pub async fn health_check(_req: Request) -> StatusCode {
    println!("{}", env::var("DATABASE_URL").expect("DATABASE_URL not set"));
    StatusCode::OK
}
```

## Mode-Specific Env Files

> System environment variables take priority, and are never overridden. After that, precedence works in order of specificity, with the most specific env file taking priority.

Reviewer 2: ```suggestion
System environment variables take priority, and are never overridden. 
After that, precedence is determined by specificity, with the most specific `.env` file taking priority.
```

```text
.env                # loaded in all cases
.env.local          # loaded in all cases, ignored by git
.env.[mode]         # only loaded in specified mode
.env.[mode].local   # only loaded in specified mode, ignored by git
```

## Caveats, Gotchas and Notes

- Secrets should only be stored in `*.local` files, as those are the only files ignored by git.

> - When you create a new env file, it is not automatically loaded. You will have to restart the server to achieve this.

Reviewer 2: ```suggestion
- When you create a new `.env` file, it is **not** automatically loaded. 
   You will need to restart the server for the changes to take effect.
```
